### PR TITLE
Add charm and bundle contents to the blob store.

### PR DESCRIPTION
**Pull request by frankban**

### CODE REVIEW

File: internal/blobstore/blobstore.go [hunk 1]
```diff
@@ -49,6 +51,15 @@ type ContentChallengeResponse struct {
...
 }
 
+func getTrustedChallengeResponse() *ContentChallengeResponse {
+	rand.Seed(time.Now().UnixNano())
+	return &ContentChallengeResponse{strconv.Itoa(rand.Int()), strconv.Itoa(rand.Int())}
+}
+
+// Trusted is a challenge response we trust.
+// When a trusted proof is provided, the associated content is always uploaded.
+var Trusted = getTrustedChallengeResponse()
```
Comment (rogpeppe) [inline internal/blobstore/blobstore.go]: just

```
var Trusted = new(ContentChallengeResponse)
```

should be fine here. It's just a token - we don't care about the contents.

File: internal/charmstore/archive.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	io.ReadSeeker
+}
+
+func (nopCloser) Close() error {
+	return nil
+}
+
+// NopCloser returns a blobstore.ReadSeekCloser with a no-op Close method
+// wrapping the provided ReadSeeker r.
+func NopCloser(r io.ReadSeeker) blobstore.ReadSeekCloser {
```
Comment (rogpeppe) [inline internal/charmstore/archive.go]: Let's not export this.

Comment (rogpeppe) [inline internal/charmstore/archive.go]: Also, better at the end, so it's not the first thing you see when looking at the file.

File: internal/charmstore/archive.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	"github.com/juju/charmstore/internal/blobstore"
+)
+
+type archiverTo interface {
+	ArchiveTo(io.Writer) error
+}
+
+// getArchive is used to turn the current charm and bundle implementations
+// into ReadSeekClosers for the corresponding archive.
```
Comment (rogpeppe) [inline internal/charmstore/archive.go]: s/the/their/
?

File: internal/charmstore/archive.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+func getArchive(c interface{}) (blobstore.ReadSeekCloser, error) {
+	var path string
+	switch c := c.(type) {
+	case archiverTo:
+		// E.g. charm.CharmDir or charm.BundleDir.
```
Comment (rogpeppe) [inline internal/charmstore/archive.go]: s/E.g./For example: /

(better to avoid latinisms, I think)

File: internal/charmstore/store.go [hunk 5]
```diff
@@ -5,38 +5,70 @@ package charmstore
...
 	"github.com/juju/charmstore/internal/router"
 	"github.com/juju/charmstore/params"
 )
 
-// Store represents the underlying charm store data store.
+// Store represents the underlying charm and blob data stores.
 type Store struct {
-	DB StoreDatabase
+	DB     StoreDatabase
+	BlobDB *blobstore.Store
```
Comment (rogpeppe) [inline internal/charmstore/store.go]: BlobStore ?

File: internal/charmstore/store_test.go [hunk 6]
```diff
@@ -33,11 +38,16 @@ func (s *StoreSuite) TestAddCharm(c *gc.C) {
 	var doc mongodoc.Entity
 	err = store.DB.Entities().FindId("cs:precise/wordpress-23").One(&doc)
 	c.Assert(err, gc.IsNil)
+
+	// The entity doc has been correctly added to the mongo collection.
```
Comment (rogpeppe) [inline internal/charmstore/store_test.go:42]: // Check that the entity doc has been correctly added to the mongo collection.
?

File: internal/charmstore/store_test.go [hunk 7]
```diff
@@ -33,11 +38,16 @@ func (s *StoreSuite) TestAddCharm(c *gc.C) {
 	var doc mongodoc.Entity
 	err = store.DB.Entities().FindId("cs:precise/wordpress-23").One(&doc)
 	c.Assert(err, gc.IsNil)
+
```
Comment (rogpeppe) [inline internal/charmstore/store_test.go:41]: We test that we can add a CharmDir here, but not a CharmArchive.
Similarly for bundles below. Perhaps make this test into a function
that accepts a Charm, so you can run the same logic both with wordpress
the charm dir and wordpress the charm archive.

**Comment by rogpeppe** [comment]
LGTM, thanks, with a few trivials and a little extra test coverage.

File: internal/charmstore/archive.go (hunk 3 above)
Comment (frankban) [inline internal/charmstore/archive.go]: Done.

File: internal/charmstore/archive.go (hunk 4 above)
Comment (frankban) [inline internal/charmstore/archive.go]: Done, but why do you think it's better to avoid latinisms?

File: internal/charmstore/store.go (hunk 5 above)
Comment (frankban) [inline internal/charmstore/store.go]: Done.

File: internal/charmstore/store_test.go (hunk 6 above)
Comment (frankban) [inline internal/charmstore/store_test.go:42]: I usually avoid using "check", "ensure", "assert" or similar in tests. We are already in the tests context, the fact we are checking something should be already clear enough.

File: internal/charmstore/store_test.go (hunk 7 above)
Comment (frankban) [inline internal/charmstore/store_test.go:41]: Fixed.

**Comment by frankban** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**